### DISCUSSION TITLE
A big part of FCS analysis time is spent in GC

### DISCUSSION THREAD

**Discussion by auduchinok** [dotnet/fsharp#13039] [General]
I'm looking at various perf snapshots and seeing that in many of them about 20-30% of the time that FCS takes to analyse a project graph is spent in GC. Here's the last example, GC takes 41 seconds:

<img width="1920" alt="Screenshot 2022-04-21 at 10 41 16" src="https://user-images.githubusercontent.com/3923587/164404844-68b5a991-1f0a-4005-a39d-d770e07fa951.png">

Could we work on finding a systematic approach that would allow us to reduce allocations where possible? Possible things are:
* reuse some collection objects
* don't create intermediate lists and use state machine list builders instead
* use arrays and mutation
* don't recalculate things where possible

Some of these things aren't very idiomatic in a functional language, but when talking about compiler and editor analysis engine, I think a better performance should matter more than being completely idiomatic.

**Comment by dsyme** [comment]
I'd encourage a focus on performance, not allocations. Allocations happen, they are not necessarily bad.   Often the gains come from avoiding running code at all more than from removing allocations

Historically the biggest gains by @tihan have been from removing transient LOH allocations.  Those are really important to find and nail down.

**Comment by auduchinok** [comment]
> I'd encourage a focus on performance, not allocations. Allocations happen, they are not necessarily bad.

I agree that performance in general should be a high priority. Allocations aren't bad until there're too many of them. We often see about 30% of time spent on GC and that does indeed look bad. It's tens seconds to minutes that our users have to wait for analysis to finish in an IDE.

> Often the gains come from avoiding running code at all more than from removing allocations

Yes, that would improve analysis time, but if the remaining code still allocates too much, it'd still spend a considerable amount of time doing GC. I think we should tackle both of these things.

@dsyme I agree that in most cases allocations aren't that bad, but in case of FCS analysis we'd like to cut down every wasted second as we expect the analysis to be as fast as possible (as our users do as well). Could you reconsider the importance of doing less allocations? I think we should try to not ignore this problem.

**Comment by auduchinok** [comment]
Here's another recent snapshot, where GC spent **26%, about 6 seconds out of ~22 seconds**::

<img width="1440" alt="Screenshot 2022-04-25 at 10 45 01" src="https://user-images.githubusercontent.com/3923587/165043616-03950419-3da2-4102-ab46-283f42d84627.png">

It looks similar to this in the most of snapshots we look at.

**Reply by TIHan** [reply]
I think a decent rule of thumb would be to get the GC to take about 15% of the time. There are still plenty of opportunities to improve here.

Last I checked, though it has been a while, GC started to take more time in Gen1.

**Comment by En3Tho** [comment]
One thing that I'd like to add is that I believe node should move away from async to some custom lightweight statemachine type. I tried to use a funky () -> Task<'a> type instead of async and it really improved compiler stacks eg. from this:

![photo_2022-04-22_20-10-46](https://user-images.githubusercontent.com/37334640/165246831-00e0cb45-fb48-4b91-b3be-bf390006d4ea.jpg)

 to this:

![photo_2022-04-23_22-29-32](https://user-images.githubusercontent.com/37334640/165246860-46d8e905-64f9-4f7e-b99c-69edbf68a48e.jpg)

It's impossble to see what is going on because of async's bindings everywhere.

I couldn't get this () -> Task thing to work properly tho because of CancellationTokens and async aware behaviour of existing code. But I believe custom type should fix this problem.

Also, there are places where F# is emitting ETW events and even tho event could be skipped, there is still string concatenation and allocations.

Many of the allocations in compiler come from Lists, Tuples and Lambdas/Closures. Maybe a custom allocation-less seq builder can help. Also, array and list equality using core "=" operator also allocates. A custom equality operator for internal usage might be better.

**Comment by dsyme** [comment]
@auduchinok The point I'm making is that it's not allocations that matter, it's performance. If we find places where reducing allocations causes better performance that's great.  But don't take the approach that reducing allocations is a good in itself - there's a long history of trying to do that in the compiler and it rarely gave measurable benefits (because Gen1 allocations and GC are rarely a bottleneck problem), and plenty of times the allocation reductions led to either more complicated code or more copying or less data-structure-sharing. 

LOH allocations were a significant exception to this rule, plus of course "stupid" allocations in tight loops,

@En3Tho Yes agreed.  https://github.com/TheAngryByrd/IcedTasks shows how to define async-like cold-start tasks that pass cancellation tokens explicitly.  Likewise synchronous code that passes cancellation tokens.  We should make an internal copy of this that gives the better debugging for async/cancellable code.

**Comment by charlesroddie** [comment]
On (imm)arrays:
- https://github.com/dotnet/fsharp/pull/9259 started to use arrays but the consensus was to wait for ImmutableArray.
- IMO it would be good to replace most lists with ImmutableArray unless the linked list structure is specifically advantageous.
- I started to do `[: :]` syntax for ImmutableArray which I think would unblock this (as you just need to add dots). Eventually I will get round to finishing this but if someone has 30mins to show me around the compiler it would speed things up a lot!

Other:
- Notes on span by @cartermp https://github.com/fsharp/fslang-suggestions/issues/887#issuecomment-661598681 . Larger project.
- Struct unions/enums, struct tuples

I don't agree with only looking for bad cases and then optimizing those. I think there are a lot of classes of performance hygiene which do not increase complexity or result in significantly more copying and less sharing.

**Comment by safesparrow** [comment]
I agree with the sentiment @auduchinok and others present.

While I completely agree that allocations per se are nothing bad, I think everyone who mentions the goal of reducing allocations means speeding up overall performance by reducing the total time blocked by GC.

As mentioned in https://github.com/dotnet/fsharp/discussions/12526 I think working on a framework&CI for benchmarking the FSharp.Compiler.Service codebase would be very useful mid- and long-term. It would help support these kind of discussions where there is no obvious answer and back them with easily-available objective data.
As all the talks on performance I've seen so far start with, "measure first".

Naively and without any results to prove it, I would think that using more optimized (time and memory wise) data structures widely in the codebase could have the following benefits:
- reduced GC time
- faster operations (eg. `List.map vs Array.map`)
- more condensed data = more items can fit in CPU caches at the same time = profit (I have no idea what that profit might be though)

**Reply by vzarytovskii** [reply]
I think one of the points which is missing l, but not less important - during our short investigation with @baronfel we noticed that we have many objects (for example TType, EntityRef instances) being promoted to Gen2. We need to understand why, and if it's necessary.

**Reply by baronfel** [reply]
For folks that want to get qualitative data on this, I can highly recommend launching Ionide and attaching to the underlying FsAutoComplete process as you mess around in the editor. We use a lot of the FCS APIs directly (not just checking, but also completions, parsing, structure, semantic analysis) and a lot of these are triggered in the background for you as you simply navigate around your process, so taking a dump of just that one process gives you a nice, isolated snapshot of the APIs in a real world scenario. That's what I did with Vlad and it was quickly apparent that the gen2 heaps were quite large.

**Reply by safesparrow** [reply]
@baronfel 
I've done something similar although more isolated: take a dump of the inputs that Rider sends to parse and check an fs file (`FSharpChecker.ParseAndCheckFileInProject`), which I would then profile in an isolated unit test.

However:
1. It required quite a bit of time to write custom JSON serialization code
2. The serialized data is specific to my workstation - I'm trying to write some code that would generalise the captured inputs, along with code that will fetch a specific codebase at a specific revision and run the analysis. Perhaps better to do via containers, but I don't have much experience with them.
3. These benchmarks and results should be readily available for everyone to run, and they should be run automatically in the CI as well and reports of those generated and stored on CI & linked from this repo. This would lower the bar for investigating performance improvements, make it much easier to objectively argue about performance on PRs, provide history of performance in each F# version, and more.

An example issue where having objective benchmark results auto-generated (or at least easy to generate by the user) on the PR would massively help the discussion which otherwise lacked any hard data: https://github.com/dotnet/fsharp/pull/9259 . There are more examples like this.

**Reply by vzarytovskii** [reply]
@safesparrow 

> they should be run automatically in the CI as well and reports of those generated and stored on CI & linked from this repo

This one is the most difficult by far. We need to make sure we use the same infrastructure every time we run benchmarks in CI, so they will be comparable between runs (same hardware ideally).

One thing we can do is to reuse infrastructure `dotnet/performance` provides. In this case we won't be able to run tests on each PR, since, i believe they only have limited amount of machines they run their tests on)

And instead we should do something smarter (either on-demand runs or once every N days).

**Reply by safesparrow** [reply]
EDIT: I agree that running something like this on every commit on every branch is too much and unnecessary. If there was a button on a PR that would trigger this on demand that would be great. Not sure how this works with forks - I assume such a button would only be available for branches on this repo? And how to stop people from abusing it?
If it's also easy to run locally, people would be doing that anyway even before they raise PRs - that would drive their research and once they get some interesting results they could raise a PR with a batch of changes and push that button once.

@vzarytovskii 
> This one is the most difficult by far

I think this is doable with cloud infra. Disregarding cost for a second, how hard is it to integrate something like this on GitHub/Azure pipeline this codebase is using atm? (genuine question)
I would think repeatable hardware is quite easy to get these days, although might be more costly.
Allocations & IO don't really require identical hardware, but CPU does as you say.

It might be useful to come up with some numbers.
In the last 19 days there were 35 commits to the main repo and 35 PRs merged.
Let's assume 2 commits on `main` per day.
Parsing/checking Fantomas for the first time in Rider takes ~12s on my machine, takes 300MB of memory and 4GB of memory allocations. It's mostly single-threaded but let's assume 12s * 2 cores = 24 CPU core seconds. Let's assume we run 20 such tests (although I think one would already provide huge value), that's 8 CPU core minutes.
There is also need to fetch&extract some code from GitHub (probably best to download a ready to analyse artefact, either already built or not).
Let's say the whole test takes 10 CPU core minutes. That's `10*2*365min ~ 122h` of CPU compute annually + memory + network costs.

If the issue turns out to not be technical difficulty but cost, I could ask whether my company would be willing to sponsor some of it.
If it is indeed technical difficulty, I would be happy to try & help with the efforts 👍

**Reply by vzarytovskii** [reply]
@safesparrow 

> how hard is it to integrate something like this on GitHub/Azure pipeline this codebase is using atm? (genuine question)

It will be easy to add a separate leg to run and publish BDN results right now. However, i don't think we have control over which images we will be running on (it may be a different VM for each run). And we can only run on certain pools due to the policies we have in place.

We will solve it, but we will have to take a bunch of things into consideration, research existing infrastructure (which other .net components are using, etc).

As I said in the separate comment, I will publish more details next week, so please, keep comments and suggestions coming!

**Reply by cartermp** [reply]
I would caution against the route of having benchmarks running in CI and using them as the basis for performance investigations. If doing many of these investigations and fixing some of them has taught me anything, it's that a lab environment and creative benchmarks will simply not capture what actually happens out in the wild. It's signing people up for a lot of work that will end up not being representative of actual user interactions on a variety of machines.

**Reply by safesparrow** [reply]
I strongly disagree. What alternative is there? The only alternative I saw is contributors attaching a profiler to their FCS service, performing an undefined sequence of operations and posting brief summaries - it provides a single data point with undefined inputs, rather than a set of well-known ones. Which is the same but less reliable, less accurate, more effort and therefore discouraged.

I think it's quite the opposite - not having those benchmarks is exactly what generates a lot of redundant work. I linked to an example PR where a lot of work was done to try&improve performance, but reviewers rightly questioned the utility since there was no data to confirm any improvements.

I realize there is no one sequence of inputs that is the only important one - that's why I think the set of benchmarks should be expanded in time as new examples are found that provide more insight.

Also the example inputs I had in mind were very high-level - doing exactly the same API call that Rider does when I open a file in Fantomas. That single call takes 12s on my machine and I really think that benchmark run on CI or other devs' machine would be very representative.

What do you think makes having ready to run benchmarks worse than the current state of affairs?

**Reply by cartermp** [reply]
The main challenge is a bit of a conundrum. To have a reliable check in CI, you want a lab environment so that machine variation doesn't impact the results. But that's exactly what makes the benchmarks unrepresentative, because user machines vary greatly. I've observed several times that my performance issues are not the same as others even when working in the same codebase, to the degree of 5-10% more or less CPU time to compute things or 20% more time spent in GC for a given sequence of inputs.

Benchmarks themselves aren't inherently bad, and we've put them to good use here: https://github.com/dotnet/fsharp/blob/main/tests/benchmarks/CompilerServiceBenchmarks/Benchmarks.fs

There are some compiler APIs you can directly validate on your machine and I think folks would welcome adding more, especially when looking into a specific API to improve the speed of. But getting into the realm of automating them is fraught with problems. This is not a one-and-done deal and would mean a pretty significant amount of time, over time, spent twiddling things, or risk the benchmarks no longer being useful in CI anymore.

Ultimately, this is an economic endeavor, I think that it's a lot more economical to just do stuff by hand right now.

**Reply by safesparrow** [reply]
I would argue that if benchmarks on the CI are discouraged because they don't represent all users' experience, then an individual dev running those benchmarks represents all users' experience even less so.

If running benchmarks locally is encouraged when working on performance PRs, and contributors are encouraged to present results on PRs to argue their case, I don't see how running the same thing on a baseline hardware is worse.
Why should the particular contributor's perf experience, who happened to run the benchmark with the change they've just made, be more valuable than all the others' who didn't?

Plus, as mentioned, with automated benchmarks one can add a plethora of hardware configurations, simulate busy machines etc - something that's simply not possible to do manually. If the concern is about performance analysis that considers a variety of configurations, then that's the reason _for_ automated benchmarks, including CI benchmarks - not against it.

I find the description in https://github.com/dotnet/performance/issues/2457 very close to my thinking.
Especially the "Approximate needs" section bit:
>Execution:
> - Run regularly on SDK releases and builds
> - Run regularly where the SDK draws F# compiler tooling from the main and release branches on dotnet/fsharp
> - Manually request a run where the SDK draws F# compiler tooling from PRs in dotnet/fsharp

RE: Existing benchmarks
It would be nice to reach a point where running all the existing benchmarks and gathering/exporting results is as simple as running the `build.cmd` commandline. If that's already the case, then I probably missed something.

**Reply by cartermp** [reply]
As I said, it's an economic endeavor. I think it's more work to build and maintain the system than it is worth. This is just based on my experience being one of the main contributors to performance improvements of FCS.

**Reply by safesparrow** [reply]
Whether building the system is worth the effort is one thing. However, what was argued later on was that the changes would have negative effect on the codebase and future work and I still don't agree with that.

**Reply by safesparrow** [reply]
@cartermp I think I misunderstood some of your original comment, sorry about that.

If the main concern is the cost of building the system, then:
1. The CI element of it is the final step. Making it easier to add, perform and export benchmark results in a consistent way, even if that's run on local boxes, has a lot of value IMO. More documentation and guides would help.
2. I think a few people, including myself, would be happy to help with the efforts.

Thanks for the discussion!

**Comment by vzarytovskii** [comment]
Here's the issue in performance repo https://github.com/dotnet/performance/issues/2457
